python/unityagents/environment.py

```diff
@@ -509,11 +366,95 @@ def close(self):
         """
         Sends a shutdown signal to the unity environment, and closes the socket connection.
         """
-        if self._loaded & self._open_socket:
-            self._conn.send(b"EXIT")
-            self._conn.close()
-        if self._open_socket:
-            self._socket.close()
-            self._loaded = False
+        if self._loaded:
+            # if self.proc1 is not None:
+            #     self.proc1.kill()
+            self.communicator.close()
         else:
             raise UnityEnvironmentException("No Unity environment is loaded.")
+
+    @staticmethod
+    def _flatten(arr):
+        """
+        Converts dictionary of arrays to list for transmission over socket.
+        :param arr: numpy vector.
+        :return: flattened list.
+        """
+        if isinstance(arr, (int, np.int_, float, np.float_)):
+            arr = [float(arr)]
+        if isinstance(arr, np.ndarray):
+            arr = arr.tolist()
+        if len(arr) == 0:
+            return arr
+        if isinstance(arr[0], np.ndarray):
+            arr = [item for sublist in arr for item in sublist.tolist()]
+        if isinstance(arr[0], list):
+            arr = [item for sublist in arr for item in sublist]
+        arr = [float(x) for x in arr]
+        return arr
+
+    @staticmethod
+    def _bytes_to_array(image_bytes, gray_scale):
```

Comment:
Comment: what, why?

Reply:
Will rename to _process_pixels to be true to the current code